### PR TITLE
EDITIONS: hook up manage editions frontend to fake API 

### PR DESCRIPTION
**Pull request by ajwl**
## What's changed?
In order to get ready for the backend of editions, we have mocked out an API and connected the frontend up to it. 

Three methods:
* GET: `fetchIssueByDate`
* GET: `fetchIssuesForDateRange`
* POST: `createIssue`

One of these - `fetchIssuesForDateRange` will not be available in the first pass of the API. Some visual features won't work because of this (ie. the highlighting of dates on the calendar), but the basic user journey should still work. 

Journey can be found here: https://fronts.local.dev-gutools.co.uk/v2/manage-editions

## Implementation notes

To mock out the backend we set up an express server with 3 endpoints.

### The Fake API - express server 

<details><summary>Click to view the index.js of the basic express server we used to mock backend</summary>
<p>

```javascript
const express = require("express");
const moment = require("moment");
const cors = require("cors");
const app = express();
const port = 3000;

app.get("/", (req, res) => res.send("Hello World!"));

app.use(cors());
app.use(express.json());
app.use(express.urlencoded({ extended: true }));

const collection1 = {
  id: "collection1",
  name: "Collection 1",
  prefill: "Capi pls",
  isHidden: false,
  updatedOn: "2019-06-29 19:10:25",
  updatedBy: "Jon and Anna!",
  updatedEmail: "jonathon.herbert@guardian.co.uk",
  live: [],
  draft: []
};

const front1 = {
  id: "front1",
  name: "Front 1",
  isHidden: false,
  updatedOn: "2019-06-29 19:10:25",
  updatedBy: "Jon and Anna!",
  updatedEmail: "jonathon.herbert@guardian.co.uk",
  collections: [collection1]
};

const issue1 = {
  id: "issue1",
  name: "Issue 1",
  publishDate: "2019-06-29 19:10:25",
  createdOn: "2019-06-29 19:10:25",
  createdBy: "Jon and Anna!",
  createdEmail: "jonathon.herbert@guardian.co.uk",
  fronts: [front1]
};

const issue2 = {
  id: "issue2",
  name: "Issue 2",
  publishDate: "2019-06-30 19:10:25",
  createdOn: "2019-06-30 19:10:25",
  createdBy: "Jon and Anna!",
  createdEmail: "jonathon.herbert@guardian.co.uk",
  fronts: [front1]
};

const issue3 = {
  id: "issue3",
  name: "Issue 3",
  publishDate: "2019-07-01 19:10:25",
  createdOn: "2019-07-01 19:10:25",
  createdBy: "Jon and Anna!",
  createdEmail: "jonathon.herbert@guardian.co.uk",
  fronts: [front1]
};

const issueTemplate = (date, id) => {
  return {
    id: `issue${id}`,
    name: `Issue ${id}`,
    publishDate: `${date}`,
    createdOn: `${date}`,
    createdBy: "The robots!",
    createdEmail: "jonathon.herbert@guardian.co.uk",
    fronts: [front1]
  };
};

// GET single issue - for date
app.get("/editions-api/issues/:date", (req, res) => {
  res.set("Access-Control-Allow-Origin", "*");
  const date = req.params.date;
  const matchingIssue = [issue1, issue2, issue3].find(i =>
    moment(i.publishDate).isSame(date, "day")
  );
  if (matchingIssue !== undefined) {
    res.status(200);
    res.json(matchingIssue);
  } else {
    res.status(404);
    res.send("No issue found");
  }
});

// GET issues - by date range
app.get("/editions-api/issues", (req, res) => {
  res.set("Access-Control-Allow-Origin", "*");
  res.json([issue1, issue2, issue3]);
});

// POST - create new issue
app.post("/editions-api/issues", (req, res) => {
  res.set("Access-Control-Allow-Origin", "*");
  const date = req.body.date;
  const id = parseInt(Math.random() * 100); //generate fake id
  const issue = issueTemplate(date, id);
  res.json(issue);
});

app.listen(port, () => console.log(`Example app listening on port ${port}!`));

```

</p>
</details>

### General
- [ ] 🤖 Relevant tests added
- [ ] ✅ CI checks / tests run locally
- [ ] 🔍 Checked on CODE

### Client
- [ ] 🚫 No obvious console errors on the client (i.e. React dev mode errors)
- [ ] 🎛️ No regressions with existing user interactions (i.e. all existing buttons, inputs etc. work)
- [ ] 📷 Screenshots / GIFs of relevant UI changes included

### CODE REVIEW

File: client-v2/src/services/editionsApi.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    {
+      method: 'get',
+      credentials: 'same-origin'
+    }
+  )
+    .then(response => {
+      if (response.status === 200) {
+        return response.json();
+      } else {
+        return [];
```
Comment (ajwl) [inline client-v2/src/services/editionsApi.ts]: Is this else statement needed? Did we determine that the catch would be fired if the if statement fails?

File: client-v2/src/services/editionsApi.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import { EditionIssue } from 'types/Edition';
+import { Moment } from 'moment';
+import pandaFetch from './pandaFetch';
+
+const dateFormat = 'YYYY-MM-DD';
+
+export const fetchIssuesForDateRange = async (
```
Comment (Reettaphant) [inline client-v2/src/services/editionsApi.ts]: This endpoint while useful is not going to be part of the editions first pass.

Comment (ajwl) [reply client-v2/src/services/editionsApi.ts]: Yes indeed, I refer to that in the writeup. We decided to leave in here but can remove if you feel strongly

File: client-v2/src/services/editionsApi.ts (hunk 1 above)
Comment (jonathonherbert) [reply client-v2/src/services/editionsApi.ts]: Now that I look at this code, I'm quite surprised that this compiles (possibly pandaFetch's typing isn't too strict) -- it feels like the return type should be `Promise<EditionIssue|null>` or similar. We definitely don't need the else, and we could choose to explicitly return null after the if condition,  or simply write nothing there to return `undefined`.

**Pull request merged**